pkg/ottl/functions.go

```diff
@@ -772,3 +803,94 @@ func NewTestingOptional[T any](val T) Optional[T] {
 		hasValue: true,
 	}
 }
+
+// typedGetter is like Getter, but with typed return values.
+type typedGetter[K, V any] interface {
+	Get(ctx context.Context, tCtx K) (V, error)
+}
+
+// Ensure LiteralGetter implements typedValueWrapper.
+var _ typedValueWrapper = (*LiteralGetter[any, any, typedGetter[any, any]])(nil)
+
+// LiteralGetter allows OTTL functions to use getters that might return literal
+// values. It provides a way to check if the getter is a literal getter and to retrieve
+// the literal value from it.
+// K is the type of the context, G is the type of the getter, and V is the type of
+// the literal value returned by getter G.
+type LiteralGetter[K, V any, G typedGetter[K, V]] struct {
+	getter G
+}
+
+func (*LiteralGetter[K, V, G]) getWrappedType() reflect.Type {
+	return reflect.TypeFor[G]()
+}
+
+func (p *LiteralGetter[K, V, G]) setWrappedValue(val reflect.Value) error {
+	typedVal, ok := val.Interface().(G)
+	if !ok {
+		return fmt.Errorf("cannot set value of type %s to a Getter of type %s", val.Type(), reflect.TypeFor[G]())
+	}
+	p.getter = typedVal
+	return nil
+}
+
+func (p *LiteralGetter[K, V, G]) getRawValue() any {
+	return p.getter
+}
+
+// IsLiteral checks if the wrapped getter holds a literal getter.
+func (p *LiteralGetter[K, G, V]) IsLiteral() bool {
+	lg, ok := any(p.getter).(literalGetter)
+	return ok && lg.isLiteral()
+}
+
+// GetLiteral retrieves the literal value from the getter.
+// If the getter is not a literal getter, or the value it's holding is not a literal,
+// an error is returned.
+func (p *LiteralGetter[K, V, G]) GetLiteral() (V, error) {
```

Comment:
Im a bit confused by this implementation. My current understanding is that we're trying to let a function like `IsMatch` take both a literal string as the input or a path. Then, in the factory of the function, let the function check if the input is a literal and if it is, efficiently make the regex once now. If it is not, allow the function to call the dynamic getter to make the regex on every call through the function.

This implementation looks like if the value isn't a literal it will error, so I dont see why I'd use it over the existing `literalGetter` struct.

Reply:
Sorry for the super long delay. As discussed offline, I think `ottl.GetLiteralValue` that returns `(val any, ok bool)` is the cleanest solution here. It makes it clear to the function author that they are retrieving a value in a static context in a setting where the underlying Getter doesn't support literal values or just in this case doesn't have one, while not burdening them with the details of how the literal value is retrieved.